src/Generator/Generators/CSharp/CSharpTextTemplate.cs

```diff
@@ -196,6 +198,138 @@ public void GenerateHeader()
+                    callingConvention != ds.callingConvention)
+                {
+                    return false;
+                }
+                return true;
+            }
+        }
+
+        private string GenerateDelegateSignature(Function function, out CSharpTypePrinterResult retType)
+        {
+            var @params = new List<string>();
+
+            TypePrinter.PushContext(CSharpTypePrinterContextKind.Native);
+
+            var retParam = new Parameter { QualifiedType = function.ReturnType };
+            retType = retParam.CSharpType(TypePrinter);
+
+            var method = function as Method;
+            var isInstanceMethod = method != null && !method.IsStatic;
+
+            if (isInstanceMethod && Options.IsMicrosoftAbi)
+            {
+                @params.Add("global::System.IntPtr instance");
+
+                if (method.IsConstructor)
+                    retType = "global::System.IntPtr";
+            }
+
+            if (!function.HasIndirectReturnTypeParameter &&
+                isInstanceMethod && Options.IsItaniumLikeAbi)
+                @params.Add("global::System.IntPtr instance");
+
+            int i = 0;
+            foreach (var param in function.Parameters)
+            {
+                if (param.Kind == ParameterKind.OperatorParameter)
+                    continue;
+
+                var typeName = param.CSharpType(TypePrinter);
+
+                @params.Add(string.Format("{0} param{1}", typeName, i++));
+
+                if (param.Kind == ParameterKind.IndirectReturnType &&
+                    isInstanceMethod && Options.IsItaniumLikeAbi)
+                    @params.Add("global::System.IntPtr instance");
+            }
+
+            if (method != null && method.IsConstructor)
+            {
+                var @class = (Class)method.Namespace;
+                if (Options.IsMicrosoftAbi && @class.Layout.HasVirtualBases)
+                    @params.Add("int " + GeneratedIdentifier("forBases"));
+            }
+
+            TypePrinter.PopContext();
+
+            return string.Join(", ", @params);
+        }
+
+        private Dictionary<DelegateSignature, string> AllDelegates = new Dictionary<DelegateSignature, string>();
+
+        private string GenerateDelegate(Method func, bool suppressWarning = true)
+        {
+            DelegateSignature ds = new DelegateSignature();
+            ds.callingConvention = func.CallingConvention.ToInteropCallConv();
+            CSharpTypePrinterResult retType;
+            ds.signature = GenerateDelegateSignature(func, out retType);
+            ds.retType = retType.ToString();
+            ds.suppressWarning = suppressWarning;
+
+            string delegateName;
+            if (AllDelegates.TryGetValue(ds, out delegateName))
+                return "DelegatesStorage." + delegateName;
+
+            if (ds.retType.Equals("void"))
+                delegateName = "Func_" + AllDelegates.Count;
```

Comment:
Such names would result in a terrible API: Func_1, Action_2, etc. Instead, please use the signatures in the name, similar to what we had so far. Func_int_MyClass_double, Action_float_MyOtherClass, etc. As per the Mono/.NET convention, in the case of Func put the return type first and the parameters later.